library/src/blas2/rocblas_hpmv.hpp

```diff
@@ -0,0 +1,225 @@
+            //                              col-1
+            // For lower matrices, index = sigma(n-i) + row
+            //                              i=0
+            int index = upper ? ((ind_y * (ind_y + 1)) / 2) + ind_x
+                              : ((ind_y * (2 * n - ind_y + 1)) / 2) + (ind_x - ind_y);
+
+            res_A += (ind_x == ind_y ? std::real(AP[index]) : CONJ ? conj(AP[index]) : (AP[index]))
+                     * x[col * incx];
+        }
+    }
+
+    // Store partial sums for the diagonal
+    sdata[tx + ty * DIM_X] = res_A;
+    __syncthreads();
+
+    thread_id = hipThreadIdx_x + hipThreadIdx_y * hipBlockDim_x;
+    ind       = hipBlockIdx_x * DIM_X + thread_id;
+    if(thread_id < DIM_X && ind < n)
+    {
+        // Add the partial sums of each diagonal and store
+        for(rocblas_int i = 1; i < DIM_Y; i++)
+        {
+            sdata[thread_id] += sdata[thread_id + DIM_X * i];
+        }
+
+        // Update y.
+        if(ind < n)
+        {
+            if(beta != 0)
+            {
+                y[ind * incy] = (alpha * sdata[thread_id]) + (beta * y[ind * incy]);
+            }
+            else
+            {
+                y[ind * incy] = alpha * sdata[thread_id];
+            }
+        }
+    }
+}
+
+/**
+  *  Loads pointers and launches the actual calculation kernel.
+  */
+template <rocblas_int DIM_X, rocblas_int DIM_Y, typename U, typename V, typename W>
+__global__ void hpmv_kernel(bool           upper,
+                            rocblas_int    n,
+                            U              alphaa,
+                            V              APa,
+                            ptrdiff_t      shifta,
+                            rocblas_stride strideA,
+                            V              xa,
+                            ptrdiff_t      shiftx,
+                            rocblas_int    incx,
+                            rocblas_stride stridex,
+                            U              betaa,
+                            W              ya,
+                            ptrdiff_t      shifty,
+                            rocblas_int    incy,
+                            rocblas_stride stridey)
+{
+    rocblas_int num_threads = hipBlockDim_x * hipBlockDim_y * hipBlockDim_z;
+    if(DIM_X * DIM_Y != num_threads)
+        return; // need to launch exactly the same number of threads as template parameters indicate
+
+    auto AP = load_ptr_batch(APa, hipBlockIdx_y, shifta, strideA);
+    auto x  = load_ptr_batch(xa, hipBlockIdx_y, shiftx, stridex);
+    auto y  = load_ptr_batch(ya, hipBlockIdx_y, shifty, stridey);
+
+    auto alpha = load_scalar(alphaa);
+    auto beta  = load_scalar(betaa);
+
+    hpmv_kernel_calc<DIM_X, DIM_Y>(upper, n, alpha, AP, x, incx, beta, y, incy);
+}
+
+/**
+  *  U is always: const T* (either host or device)
+  *  V is either: const T* OR const T* const*
+  *  W is either:       T* OR       T* const*
+  */
+template <typename U, typename V, typename W>
```

Comment:
template<typename S, typename A, typename X, typename Y>
S as scalar type, 
A as a-type
X as x-type
Y as y-type

Reply:
I agree that U, V, W probably aren't the best descriptors...
I implemented a counter-proposal:
`TScal`, `TConstPtr`, `TPtr`
I think these are more descriptive that Scalar-type, a-type, etc. They also allow multiple variables to use the same type without adding confusion. Let me know what you think.

Reply:
I like that but are the Ts always the same type for this function?  Mixed precision function?

Reply:
AFAIK the Ts should always be the same, I don't see otherwise. T should always either be rocblas_float_complex or rocblas_double_complex.

Reply:
I wonder what would we use after U, V, W :)
I wrote a document on some useful definition of these types, that would allow us many traits mechanisms.
So far, with some of our abstractions of the types, 

<U,V,W> + { auto + load_ptr_batch }

some routines could even perform operations involving a batched vector and a strided batched vector (benefit of skipping T and use the auto keyword when loading the batch ptr, 99% to 100% of the use of load_ptr_batch is let to the compiler (a grep would confirm)). If underlying T types are different I guess the compiler would scream ... 

There is a lot of U,V,W in rocBLAS. 
Fundamentally, we are doing nothing explicitly with those types, maybe a few std::are_same ..? We let the compiler do. So, it doesn't not really matter from a 'functional' point of view, neither readability except that I think U,V,W is weird.
 
Otherwise, the question is how many template parameters are used.
S,A,B,C (here 4) or TScal,TConstPtr,TConstPtr,TPtr (here 3, and matrix a and b are of the same pointer types).

- naming is a nice to have
- number of template parameters decides whether or not we allow a mix between batched and strided batched vector / matrix.

Reply:
Your routine works with V = const T* (strided_batched)  and W = T * const * (batched). (?)

Reply:
That situation should never happen, right? I think it would work in that situation, though, as everything is handled in load_ptr_batch.

Reply:
Why not?... (only raising the question)

Reply:
For the mixed precision, I think the kernels form a barrier since (most of) they require everything being of the same type T.

Reply:
Well at least from our current API we don't support mixing batched and strided_batched routines together (having some variables batched and other variables strided_batched). As far as I know we aren't planning on supporting this in the future.
But if this did ever come up, I don't think many code changes would have to happen here as `load_ptr_batch` should automatically do the necessary conversions from a batched array to a single pointer either way.

Mixed precision is it's own issue, it only seems to come up with certain functions (gemm, trsm, etc.) and it doesn't seem worthwhile to have _every_ function future-proof for mixed precision, unless this is something that might have to be implemented across the project, but I haven't heard anything about this.

Reply:
Agree, only discussing the constraints as a consequence of the current type abstraction, which does not translate completely the C-API constraint.